### PR TITLE
fix: check for latest complete epoch before an iteration of state dump

### PR DESCRIPTION
**Pull request by ppca**
ND-437: basically even when the node is in progress for an epoch, also check if the latest local complete epoch is newer than the epoch it's in progress for, if so, switch to dump for the latest complete epoch.

### CODE REVIEW

File: nearcore/src/state_sync.rs [hunk 1]
```diff
@@ -241,101 +237,109 @@ async fn state_sync_dump(
                 match in_progress_data {
                     Err(error) => Err(error),
                     Ok((state_root, num_parts, sync_prev_hash)) => {
-                        let missing_parts = get_missing_part_ids_for_epoch(
+                        let new_epoch_state = check_new_epoch(
```
Comment (nikurt) [inline nearcore/src/state_sync.rs:240]: Please consider getting the current epoch_id at the beginning of the iteration, __before__ doing `match progress`. Doing it once should simplify things, in comparison to doing it in every branch of the `match` expression.

**Comment by ppca** [comment]
I do not call start_dumping() from check_latest_epoch() function in this change. check_latest_epoch()  will always return the latest epoch id and sync_hash no matter what progress the last iteration of dump made. 
The code only calls start_dumping when by comparing latest epoch id and last iteration epoch id, it can be sure that it is starting the dump of a new epoch. I did not merge the branches where a new epoch is started because I wanted to show different debug messages for each of them.

File: nearcore/src/state_sync.rs [hunk 2]
```diff
@@ -487,147 +494,28 @@ fn start_dumping(
...
 }
 
-/// Checks what is the latest complete epoch.
-/// `epoch_id` represents the last fully dumped epoch.
-fn check_new_epoch(
-    epoch_id: Option<EpochId>,
-    epoch_height: Option<EpochHeight>,
-    num_parts: Option<u64>,
+/// return epoch_id and sync_hash of the latest complete epoch available locally.
+fn check_latest_epoch(
```
Comment (nikurt) [inline nearcore/src/state_sync.rs:509]: Should this be renamed to `get_latest_epoch()` or similar?

**Review by nikurt — APPROVED** [review]

File: nearcore/src/state_sync.rs (hunk 2 above)
Comment (ppca) [reply nearcore/src/state_sync.rs:509]: sounds good

**Pull request merged**